primeqa/qg/models/qg_model.py

```diff
@@ -52,33 +61,110 @@ def tokenizer(self):
         """
         return self._tokenizer
 
+    def prune_hallucinations(self, qdicts, num_instances=5, hallucination_prop=0.25):
+        """
+            Prunes hallucinated questions using an entity lookup approach.
+        Args:
+            qdicts (list): A list of dicts, each dict contains, an answer, context and (num_instances) generated questions.
+        Returns:
+            Updated qdicts without any hallucinated questions. Each dict will have only one question now.
+        """
+        new_qdicts = []
+        reserve_hallucinated_qdict = []
+        reserve_non_hallucinated_qdict = []
+        question_words = set(['what', 'when', 'where', 'which', 'how', 'who', 'whose', 'whom'])
+        for qdict in qdicts:
+            answer = qdict['answer']
+            context = qdict.pop('context')
+            questions = qdict.pop('questions')
+            if not reserve_hallucinated_qdict:
+                qdict['question'] = questions[0]
+                reserve_hallucinated_qdict.append(qdict)
+            context_tokens = set([tok.text.lower() for tok in self.path_sampler.nlp_model.process(\
+                                context, processors='tokenize').sentences[0].tokens])
+            flag = False
+            same_questions = set([])
+            for question in questions:
+                if not question.strip() or question in same_questions: continue
+                same_questions.add(question)
+                doc = self.path_sampler.nlp_model(question)
+                hallucinated = set([])
+                for entity in doc.sentences[0].entities:
+                    entity = entity.text.lower()
+                    entity_tokens = set([etok.text.strip().lower() for etok in self.path_sampler.nlp_model.process(\
+                                        entity, processors='tokenize').sentences[0].tokens if etok.text.strip().lower() != "'s"])
+                    has_qs_words = question_words.intersection(entity_tokens)
+                    if has_qs_words: continue
+
+                    hcount = 0
+                    for e_tok in entity_tokens:
+                        if e_tok not in context_tokens:
+                            hcount += 1
+
+                    if hcount == 0: continue
+
+                    hall_quant = round(len(entity_tokens) * hallucination_prop) #no. of hall. words tolerable
+                    if hcount > hall_quant or len(entity_tokens) == hcount:
+                        hallucinated.add(entity)
+
+                if not hallucinated:
+                    qdict_copy = copy.deepcopy(qdict)
+                    qdict_copy['question'] = question
+                    if flag:
+                        reserve_non_hallucinated_qdict.append(qdict_copy)
+                    else:
+                        new_qdicts.append(qdict_copy)
+                    flag = True
+
+        if not new_qdicts:
+            return reserve_hallucinated_qdict
+        elif len(new_qdicts) < num_instances:
```

Comment:
What is happening here?

Reply:
In case most of the generated questions are hallucinated, pick some hallucinated questions so that the number of output questions equals `num_instances`.

Reply:
Can you add this explanation as a comment?

Reply:
I have added it as a comment to the code.